types/src/validator_verifier.rs

```diff
@@ -159,54 +168,114 @@ impl ValidatorVerifier {
+    ) -> Result<MultiSignature, VerifyError> {
+        let (aggregated_sig, aggregated_key) = self.generate_multi_signature(partial_signatures)?;
+        if let Some(aggregated_key) = aggregated_key {
+            aggregated_sig
+                .multi_sig()
+                .as_ref()
+                .expect("Failed to get multi signature")
+                .verify(message, &aggregated_key)
+                .map_err(|_| VerifyError::FailedToAggregateKey)?
         }
-        Ok(())
+        Ok(aggregated_sig)
     }
 
-    /// This function will try batch signature verification and falls back to normal
-    /// iterated verification if batching fails.
-    pub fn batch_verify_aggregated_signatures<T: CryptoHash + Serialize>(
+    /// This function will successfully return when at least quorum_size signatures of known authors
+    /// are successfully verified. It creates an aggregated public key using the voter bitmask passed
+    /// in the multi-signature and verifies the message passed in the multi-signature using the aggregated
+    /// public key.
+    ///
+    pub fn verify_multi_signatures<T: CryptoHash + Serialize>(
         &self,
         message: &T,
-        aggregated_signature: &BTreeMap<AccountAddress, bls12381::Signature>,
+        multi_signature: &MultiSignature,
     ) -> std::result::Result<(), VerifyError> {
-        self.check_num_of_signatures(aggregated_signature)?;
-        self.check_voting_power(aggregated_signature.keys())?;
-        let keys_and_signatures: Vec<(bls12381::PublicKey, bls12381::Signature)> =
-            aggregated_signature
-                .iter()
-                .flat_map(|(address, signature)| {
-                    let sig = signature.clone();
-                    self.get_public_key(address).map(|pub_key| (pub_key, sig))
-                })
-                .collect();
-        // Fallback is required to identify the source of the problem if batching fails.
-        if bls12381::Signature::batch_verify(message, keys_and_signatures).is_err() {
-            self.verify_aggregated_struct_signature(message, aggregated_signature)?
+        self.check_num_of_signatures(multi_signature)?;
+        self.check_voting_power(
+            multi_signature
+                .get_voter_addresses(&self.get_ordered_account_addresses_iter().collect_vec())
+                .iter(),
+        )?;
+        // Verify the optimistically aggregated signature.
+        let pub_keys_to_agg: Vec<&PublicKey> = multi_signature
+            .get_voters_bitmap()
+            .iter()
+            .zip(&self.address_to_validator_info)
+            .filter_map(|(voted, (_address, consensus_info))| {
+                if *voted {
+                    Some(&consensus_info.public_key)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        if self.quorum_voting_power == 0 {
+            // This should happen only in case of tests.
+            return Ok(());
```

Comment:
Uhm... what's going on here? There was already a check above:

```
        self.check_voting_power(
            multi_signature
                .get_voter_addresses(&self.get_ordered_account_addresses_iter().collect_vec())
                .iter(),
        )?;
```
We are returning `Ok()` without doing any actual multisignature validation...

Not sure I understand the semantics of this function.

Reply:
In this case the quorum voting power is 0, which means that the size of the quorum itself is 0, so there is no public key to aggregate. This happens only in case of tests - as otherwise, the `check_voting_power` function will error out.